src/engine/strat_engine/range_alloc.rs

```diff
@@ -43,10 +41,52 @@ impl RangeAllocator {
         for &(off, len) in ranges {
             self.check_for_overflow(off, len);
 
-            for val in *off..*off + *len {
-                let inserted = self.used.insert(val);
-                if inserted == false {
-                    panic!(format!("inserted value {} already present", val));
+            let contig_prev = self.used
+                .range(..off)
+                .rev()
+                .next()
+                .map(|(k, v)| (*k, *v))
+                .and_then(|(prev_off, prev_len)| {
+                    assert!(off >= prev_off + prev_len,
+                            "given range must not overlap previous range");
+                    if prev_off + prev_len == off {
+                        Some((prev_off, prev_len))
+                    } else {
+                        None
+                    }
+                });
+
+            let contig_next = self.used
+                .range(off..)
+                .next()
+                .map(|(k, v)| (*k, *v))
+                .and_then(|(next_off, next_len)| {
+                    assert!(off + len >= next_off,
+                            "given range must not overlap next range");
+                    if off + len == next_off {
+                        Some((next_off, next_len))
+                    } else {
+                        None
+                    }
+                });
+
+            match (contig_prev, contig_next) {
+                (None, None) => {
+                    self.used.insert(off, len);
+                }
+                (None, Some((next_off, next_len))) => {
+                    // Contig with next, make new entry
+                    self.used.insert(off, len + next_len);
+                    self.used.remove(&next_off);
+                }
+                (Some((prev_off, prev_len)), None) => {
+                    // Contig with prev, just extend prev
+                    *self.used.get_mut(&prev_off).expect("must exist") = prev_len + len;
+                }
+                (Some((prev_off, prev_len)), Some((next_off, next_len))) => {
+                    // Contig with both, remove next and extend prev
+                    self.used.remove(&next_off);
```

Comment:
An ```expect()``` here, since it must remove or something is wrong with our logic? This is pretty minor.